Libplanet/Store/IStateStore.cs

```diff
@@ -0,0 +1,76 @@
+using System;
+using System.Collections.Immutable;
+using System.Security.Cryptography;
+using Bencodex.Types;
+using Libplanet.Action;
+using Libplanet.Blockchain;
+using Libplanet.Blocks;
+
+namespace Libplanet.Store
+{
+    /// <summary>
+    /// An interface to abstract methods which handle states.
+    /// </summary>
+    public interface IStateStore
+    {
+        /// <summary>
+        /// Sets states mapped as relation <see cref="Block{T}.Hash"/> → states.
+        /// It guarantees <see cref="GetState"/> will return the same state if you passed same
+        /// <paramref name="blockHash"/> and haven't written overwrite the value, storage was
+        /// not broken.
+        /// </summary>
+        /// <param name="blockHash">The <see cref="Block{T}.Hash"/> to set states.</param>
+        /// <param name="states">The dictionary has multiple paris, state key → state.</param>
+        /// <param name="blockGetter">The getter returns the block corresponded to
+        /// <see cref="HashDigest{T}"/>–typed parameter.</param>
+        /// <typeparam name="T">An <see cref="IAction"/> type. It should match to
+        /// <paramref name="blockGetter"/>'s <see cref="Block{T}"/>–typed return value's
+        /// type parameter.</typeparam>
+        void SetStates<T>(
+            HashDigest<SHA256> blockHash,
+            IImmutableDictionary<string, IValue> states,
+            Func<HashDigest<SHA256>, Block<T>> blockGetter = null)
+            where T : IAction, new();
+
+        /// <summary>
+        /// Gets state queried by <paramref name="stateKey"/> in the point,
+        /// <paramref name="blockHash"/>.
+        /// </summary>
+        /// <param name="stateKey">The key to query state.</param>
+        /// <param name="blockHash">The <see cref="Block{T}.Hash"/> which the point to query by
+        /// <paramref name="stateKey"/> at.</param>
+        /// <param name="chainId">The <see cref="BlockChain{T}.Id"/> of wanted got.</param>
+        /// <returns>The state queried from <paramref name="blockHash"/> and
+        /// <paramref name="stateKey"/>. If it couldn't find state, returns `null`.</returns>
+        IValue GetState(
+            string stateKey,
+            HashDigest<SHA256>? blockHash = null,
+            Guid? chainId = null);
+
+        /// <summary>
+        /// Check the states corresponded to the block derived from <paramref name="blockHash"/>.
+        /// </summary>
+        /// <param name="blockHash">The <see cref="Block{T}.Hash"/> of <see cref="Block{T}"/>.
+        /// </param>
+        /// <returns>Whether to exist the states corresponded to <paramref name="blockHash"/>.
+        /// </returns>
+        bool ExistsBlockState(HashDigest<SHA256> blockHash);
```

Comment:
As we've consistently used plural “block states,” how about renaming this to `ContainsBlockStates()`?  (Also the verb “to exist” is inappropriate here, as we say “A exists in B” where it means “B contains A,” not “B exists A.”)

Reply:
You're right. Furthermore, there are already many methods using `Contains–` prefix, it seems better also for consistent. 🙏